src/renderer/geometry/shaders/mesh.vert

```diff
@@ -117,4 +117,9 @@ void main()
 #ifdef USE_INSTANCE_COLORS
     col *= instance_color;
 #endif
+
+    // *** IDs ***
+#ifdef USE_INSTANCE_ID
+    col.x = -float(gl_InstanceID);
```

Comment:
I'd like this to be explicit put in a new output variable `out uint instanceID;`. Why do we need the negation?

Reply:
This is tricky. By default, the color output already exists from all vertex shaders, which will be set to `1.0` if no color information is requested. Vertex shaders with instancing will then store the ID in the red channel (notated here as `x` to indicate that it isn't color information), but since `1.0` represents the absence of an instance ID, the actual ID is negated as to not overlap. In the material, it is then un-negated to revert the ID space to the positive numbers, making any non-instanced geometry have an instance ID of `-1.0`. This definitely isn't the easiest-to-understand way of implementing it, but it minimizes the number of geometry shader variants that are required to supply the necessary information for the material shader to function. Alternatively, the geometry shaders could always output a new variable, storing `-1.0` when `USE_INSTANCE_ID` is disabled and `gl_InstanceID` when it is enabled. Also, it would still have to be a float since MacOS doesn't like fragment shaders handling `uint`s. My thoughts on this were if it was going to be a float anyway, I might as well use the float that already exists.